### PR TITLE
Implement backup for the Server

### PR DESCRIPTION
**Pull request by webbnh**
Lost in the v0.69 -> v1.0 transition was the mechanism for making backup copies of result tarballs loaded on to the Server.  This gap is currently covered by the fact that the "pass-through" server is v0.69-based and so it makes a backup of the result before pushing it to the "index" server.  As things stand, once we upgrade the "index" server to a full-blown v1.0 and have users push results to it directly, we won't have any provision for creating backups.  This PR closes that gap.

This change adds a `pbench-backup-dir` configuration option which points to a directory to receive copies of result tarballs.  (This directory should be placed on a file system different from the archive directory (preferably on different hardware), although this can easily be effected via a mount point.)  This change adds two functions to the `IntakeBase` class which encapsulate the details of creating a backup copy and of removing it in the event of an upload failure.  Beyond that, the change is just to call the backup function and to declare the removal function as a cleanup handler during the intake process.  The rest of the changes are revisions to the unit tests for the impact of the configuration and backup changes along with two new unit tests for the two new functions.

The PR is structured as a series of commits for your reviewing pleasure:
- the changes to the Server code and configuration
- changes to address linter complaints in the relevant unit test code
- tweaks to `FakeCacheManager` to make it slightly less fake
- the changes to the unit test code for the new backup support

PBENCH-1259

### CODE REVIEW

**Review by dbutenhof — DISMISSED** [review]
Yes; yes, you are getting a `flake8` failure. Aside from that, fine...

**Review by dbutenhof — DISMISSED** [review]

File: lib/pbench/test/unit/server/test_upload.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_upload.py:1048]: Is this necessary? From server `__init__.py`, `PathLike = Union[str, bytes, os.PathLike]` ... so your added Union seems redundant. Unless `PathLike` isn't working at all, in which case the fix should go there ...

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/unit/server/test_upload.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:1048]: No, it's not necessary, it's "documentational" (I thought you might call this out....):  some of the list entries are actual strings, while other entries are `PathLike`'s.  Yes, I know, `PathLike` includes `str`, but it felt wrong to sweep the entries which _aren't_ paths into `PathLike`, even though they "fit"...so, I figured I would include them both.  (Hey, you were the one who "made" me add the inner types....  😉)

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_upload.py:1048]: I don't get the argument here. Sure, we could skip all aggregate types (like `JSONOBJECT`) and declare all the base types everywhere, but that seems silly. `PathLike` very deliberately includes `str`, so the first natural reaction is "wait, is there something wrong that requires duplicating this type? Did the author not *know* it was a duplicate? Why do we even *have* this type?" In any case, I approved, so the call is yours.

Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:1048]: Yeah, I don't know what to say.  (I suspect that this might be a case of the code providing "too much information".)  

In the author's mind, some of the list items are strings and some are paths, nevermind the fact that paths _could_ be represented as strings.  I felt like if I omitted the `str`, that would give the reader the incorrect impression that all the entries are paths.  🤷

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_upload.py:1048]: No, the idea is to concisely express that the values are "path like". You're you don't trust the type hierarchy and don't expect your readers to trust it, so why do we even bother with types? 🍤

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_upload.py:1048]: I still don't get your point of view at all. You're saying that all elements of the list are `PathLike` objects, but some are `Path` and some are `str` and some could conceivably be `bytes` (though we never actually use that). Well, cool, because that's *exactly* what `PathLike` means, and why it exists.

It's not just "redundant from the type checker utility's point of view", it's redundant from any conceivable point of view. (That is, if you can conceive of a point of view I can't, you've failed to explain it to me in a way that makes sense to me. 😆 )

Basically, you're simply saying that "because it might be of the type you're specifying, you should also include another type for clarity". Huh?

Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:1048]: Let's get concrete -- here's one value for `expected_ops`, from line 1036:
```
                [["copy", "md5", "bdir"], ["copy", "tbp", "bdir"], ["unlink", "bmd5"]],
```
So, that's a list of lists where each element in the outer list is a list of strings; however, in lines 1097-1106, the code replaces the placeholders, `"md5"`, `"tbp"`, `"bdir"`, and `"bmd5"`, with the current values of `md5`, `tbp`, `server_config.BACKUP`, and `server_config.BACKUP / md5.name`, respectively, which I believe are `Path`'s.  So, when the comparison is done at line 1107, the elements of `expected_ops` are lists whose first element is a `str` and whose subsequent elements are `Path`'s (or maybe `str`'s...I can't keep track...).  So, the type of the inner list should (in my opinion) be `list[str, Path]`, but I think that `list[str, PathLike]` captures it better (because I'm not sure what the actual path instances' types are, but they should conform to `PathLike`).

Does it make more sense, now?  😀

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_upload.py:1048]: I guess I maybe understand what you think you meant, but I still don't see any sense in it.

Part of the hacky obscurity and differing POVs here comes from the fact that you're trying to make your `expected_ops` array do double-duty: the result is "unnatural". You're trying to explicitly and separately describe an *input* of `str` and an *output* of `Path`. First that's syntactically and in language terms semantically pointless. Beyond that, modifying the actual argument array within the code is bizarre in the first place and changing the type while you're doing it is even more so.

Logically you seem to be looking for a *transformation* more like

```
    assert ops == [
        [
            {
                "tpb": tpb,
                "md5": md5,
                "bdir": server_config.BACKUP,
                "bmd5": server_config.BACKUP / md5.name
            }[x] for x in a
        ] for a in expected_ops
    ]
```

where `expected_ops: list[list[str]]` and the output (if you instantiated it with a name and type) would be `list[list[Path]]`. Modifying the input list is ugly in the first place, trying to describe both types at once is uglier and also makes it unnecessarily obscure.

But again, I approved, I'm not a big fan of how you chose to do that but "let he without hacks throw the first changes requested", and I leave it to you. 🤣

Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:1048]: I'm not trying to make my `expected_ops` do double-duty -- I'm trying to overcome the fact that the parametrize'd list cannot reference any values which are not global scope (e.g., I need to reference values from fixtures and things which are bound too late).  So, I've filled it with placeholders which I replace once I have the actual values.

The other detail that I think you're missing is that the first element in each of the inner lists is _not_ a `Path` -- it's a `str` indicating the operation, so `list[list[Path]]` isn't the right type in any case.

I'm not a fan of heterogenous lists, so it would be nice to make `expected_ops` be a `dict[str, list[Pathlike]]`, using the operation as the key, but that won't work because `"copy"` shows up more than once.  I could go with something like `list[list[str, list[Pathlike]]]` but the extra level is unpleasant.

In any case, I sort of like your suggested code (I really like the table look-up better than my cascaded `if` statement; however, making it cover the operation string within the comprehension seems a little challenging), but when a comprehension spreads across that many lines, I start to think that a `for` loop would be better....  🙂

Comment (dbutenhof) [reply lib/pbench/test/unit/server/test_upload.py:1048]: Hmm, I think that the non-placeholder here just confuses the complication rather than excusing it. But, yeah, it does complicate a comprehension. (I.e., it's hard to comprehend. 😄 ) You basically want to copy `[0]` and translate `[1:]` for each sub-list... ☹️

Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:1048]: I like the way you phrased that.  🤣

And, yes, I think you've put your finger on it.

But, I _think_ I have a solution, and the comprehension is not entirely incomprehensible.

**Comment by webbnh** [comment]
I've squashed today's commits together (because some of them were a tangent), so you may have to re-review a couple of lines (but, mostly, I got _rid_ of changes).

But, you might want to hold off until the unit tests pass....

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/unit/server/test_upload.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_upload.py:932]: Seems like doing `mock_out_backup` as a fixture would have removed the need to add `monkeypatch` where it wasn't otherwise needed; but then you'd have needed to add the backup fixture instead...

File: lib/pbench/test/unit/server/test_upload.py
Comment (dbutenhof) [inline lib/pbench/test/unit/server/test_upload.py:1116]: 👍🏻

**Review by dbutenhof — APPROVED** [review]

File: lib/pbench/test/unit/server/test_upload.py
Comment (webbnh) [reply lib/pbench/test/unit/server/test_upload.py:932]: Right:  I didn't want to make `mock_out_backup` a module-scope fixture (or other large scope), because some of the tests want to mock it out in other ways, and making it a function-scope fixture means adding it to the parameter list (instead of or in addition to `monkeypatch`).

**Pull request merged**
